src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java

```diff
@@ -650,63 +456,28 @@ static Map<Kmer, Integer>  getKmerReadCounts(final List<GATKRead> hardClippedPil
      * @return A set of artificial haplotypes limited to at most numPileupHaplotypes
      */
     @VisibleForTesting
-    static Set<Haplotype> filterPileupHaplotypes(final List<Haplotype> onlyNewHaplotypes,
-                                                 final Map<Kmer, Integer> kmerReadCounts,
+    static Set<Haplotype> filterPileupHaplotypes(final Set<Haplotype> onlyNewHaplotypes,
+                                                 final Map<Kmer, MutableInt> kmerReadCounts,
                                                  final int numPileupHaplotypes,
                                                  final int kmerSize ) {
-
-        // Get haplotypes from assemblyResultSet and kmerize. for each haplotype create a set of kmers.
-        // for each haplotype, look up the kmers in the read-map and sum the counts fo the haplotype score
-        // create a Map<Haplytope, Score>
-        LinkedHashMap<Haplotype, Integer> haplotypeScores = new LinkedHashMap<>();
-        for (Haplotype haplotype : onlyNewHaplotypes) {
-            // TODO this code might use some normalizing based on haplotype length in the future
-            Set<Kmer> hapKmers = kmerizeBytes(haplotype.getBases(), kmerSize);
-            int hapKmerCount = 0;
-            for(Kmer kmer : hapKmers) {
-                hapKmerCount += kmerReadCounts.containsKey(kmer) ? 1 : 0;
-            }
-
-            haplotypeScores.put(haplotype, hapKmerCount);
+        if (onlyNewHaplotypes.size() <= numPileupHaplotypes) {
+            return onlyNewHaplotypes;   // if the limiting haplotype count is more haplotypes than we have, everything passes
         }
 
-        // Select the top haplotypes in order of how many of their kmers have supports in the underlying reads
-        Map<Haplotype,Integer> topScoringHaplotypes =
-                haplotypeScores.entrySet().stream()
-                        .sorted(Collections.reverseOrder(Map.Entry.comparingByValue()))
-                        .limit(numPileupHaplotypes)
-                        .collect(Collectors.toMap(
-                                Map.Entry::getKey, Map.Entry::getValue, (e1, e2) -> e1, LinkedHashMap::new));
+        // sort by score = # of kmers in haplotype that appear in any read
+        // TODO this code might use some normalizing based on haplotype length in the future
```

Comment:
Honestly we should just make a ticket in github to revisit this entire codepath at some later date... i don't like this heuristic and i suspect we could do much better... thank you for not cutting too deeply here...

Reply:
also i will warn you this code WAS a performance hotspot so you might want to consider if any of these streams are going to be heavy duty.... Particularly anything looping over kmers...

Reply:
This shouldn't be a problem because it's only one stream per haplotype.  If it were streaming over every kmer (or of course if the CARROT tets showed a performance regression) I would be more worried.